CONTRIBUTING.md

```diff
@@ -59,9 +55,30 @@ If your code requires new dependencies, be sure that those are available on all
 
 ### User experience
 
-Jamulus is used by people all over the world with different backgrounds and knowledge. Features and language should be user focused.
-The Jamulus project has a [style and tone guide](https://jamulus.io/contribute/Style-and-Tone) which you should follow to remain consistent.
-Features should be usable in the sense that they act as expected to somebody who does not have a technical background.
+Jamulus is used by people all over the world with different backgrounds and knowledge. Therefore, think user focused.
+
+- Features should be usable in the sense that they act as expected to somebody who does not have a technical background.
+- Make sure you follow the [style and tone guide](https://jamulus.io/contribute/Style-and-Tone) to remain consistent.
+
+### Submitting code
+
+- Create a Github fork of the Jamulus code and create a Pull Request from that fork when you are done. Please don't use diff/patch to send your code changes. GitHub can be confusing, so feel free to ask for help. The [git-scm book](https://git-scm.com/book/en/v2) might be a good reference to get used to git.
+- Avoid submitting code from your master or main branch. This will confuse you. Instead create a feature branch on your repo via `git checkout -b feature/<insertNameOfFeatureHere>` and apply your changes there. For help on branches [see the Git book](https://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging).
+- Please run a local (build/feature) test to ensure there are no errors. After opening a pull request, keep an eye on the CI checks for quality or compile issues and fix them as required. You can also test the build on your repository by naming your branch `autobuild/<branchName>` to invoke the building process on your repo.
+
+#### Updating a PR/fixing conflicts
+
+If there are conflicts with jamulussoftware/jamulus showing when you raise your pull request or if you are requested to (rebase)[https://teamairship.com/update-outdated-branch-without-creating-merge-commit/] your work, do so as follows:
+
+```shell
+git remote add upstream git@github.com:jamulussoftware/jamulus ;# add the main repo as upstream remote if you didn’t before already to get the latest code
+git checkout master ;# checkout your local master
+git pull upstream master --tags ;# refresh from jamulussoftware/jamulus (i.e. the main repository)
+git push --tags --force ;# update your fork's master
+git checkout - ;# back to your working branch
+git rebase master ;# replay your changes onto the latest code (and fix conflicts if needed)
```

Comment:
```suggestion
git fetch upstream
git rebase upstream/master ;# replay your changes onto the latest code (and fix conflicts if needed. Afterwards continue:)
git rebase --continue
git push --force
```

Reply:
Would that be ok?

Reply:
> Would that be ok?

Definitely better than the original, but I like to keep my origin master in sync with upstream too, so I can rebase against origin (A few steps more, but I think this gives less chances on problems.). Also I like to use `git push --force-with-lease` on my branches to make sure I'm not messing things up.
So that would be:

```
# Update origin/master:
git fetch upstream
git checkout master
git reset --hard upstream/master  
git push origin master --force
```
After that you can do:
```
# Rebase your branch(es):
git checkout <theBranchYourWorkIsOn>
git rebase -i origin/master   # This should now be the same as git rebase -i upstream/master, but I'm sure origin did not change, while upstream can have changed in between when I rebase multiple branches.
...                           # Solve conflicts when neccesary...
git rebase --continue
git push --force_with_lease

```

Reply:
Yes, but I'd not like to "force" the user to update the master branch. Wasn't that what confused you at the beginning?

Reply:
I'm not sure what caused all my initial problems... But still I think we should advise the safest way, not the easiest...

Reply:
Safest = not modifying master at all (people might work on master which will start to get confusing)

Reply:
> Would that be ok?

- Why is `-i` necessary for `rebase` here?
- `git push --force` could be made a bit safer. `git push` behavior without refs depends on the git version and the specific config. `git push origin <theBranchYourWorkIsOn>` might be better. I'm a bit torn about `--force` vs. `--force-with-lease`. In general, I'd really prefer the latter, but it is only available in newer git versions and an author might rather want to deliberately overwrite their remote branch even if it was not fetched recently. So I guess I'd keep `--force`.

Reply:
`-i` isn't necessary. But maybe good if we tell the user to squash commits?

Reply:
Well but that would need more instructions on what to do anyway, so maybe keep it simpler here and go without `-i`?

Reply:
```suggestion
git fetch upstream ; # Get the latest changes from the main upstream repo you added before (if needed)
git checkout <theBranchYourWorkIsOn> ;# switch to (= checkout) the branch you want to rebase
git rebase upstream/master ;# replay (=rebase) your changes onto the latest code (and fix conflicts if needed)
git rebase --continue # to continue the rebasing progress after having fixed the conflicts
git push --force # Push the changes to your repo
```